src/Core/Settings.h

```diff
@@ -1066,6 +1066,7 @@ class IColumn;
     M(Bool, input_format_json_defaults_for_missing_elements_in_named_tuple, true, "Insert default value in named tuple element if it's missing in json object", 0) \
     M(Bool, input_format_json_throw_on_bad_escape_sequence, true, "Throw an exception if JSON string contains bad escape sequence in JSON input formats. If disabled, bad escape sequences will remain as is in the data", 0) \
     M(Bool, input_format_json_ignore_unnecessary_fields, true, "Ignore unnecessary fields and not parse them. Enabling this may not throw exceptions on json strings of invalid format or with duplicated fields", 0) \
+    M(Bool, input_format_json_infer_variant_from_multi_type_array, false, "Try to infer variant type rather than tuple when column/array has multiple", 0) \
```

Comment:
```suggestion
    M(Bool, input_format_try_infer_variants, false, "Try to infer Variant type in text formats when there are more than one possible type for column/array element", 0) \
```

Let's do it not only for JSON arrays, but for all text input formats in all cases when there are multiple types for single column.

For example in CSV format:
```csv
1,Hello World!
2,"[1,2,3]"
3,"2020-01-01"
```
We will infer columns `c1 Int64, c2 Variant(String, Array(Nullable(Int64)), Date)`.

Reply:
@Avogar sorry for the delay! I've pushed the latest changes now. At the moment having a little trouble figuring out how to properly infer variant for other formats as well. For `SELECT c1, toTypeName(c1), c2, toTypeName(c2) FROM format('CSV', '1,Hello World!\n2,"[1,2,3]"\n3,"2020-01-01"\n') FORMAT Pretty;` when I step through the code with debugger I can see it hitting the variant transformation but in the end `Nullable(String)` gets inferred still.